Signup: DSS: Add DangerousHtml component to prevent rendering errors

sirbrillig opened the pull request:
Normally it's possible to use the special `dangerouslySetInnerHTML` prop of a
component to add raw HTML into that component, but the markup is injected as-is,
which means that any invalid markup will risk breaking React's virtual DOM.
(Generally this results in errors like `Invariant Violation: Expected markup to
render 15 nodes, but rendered 10`).

DangerousHtml creates a tag (of the tag name matching `tag`) and uses the web
API's `innerHTML` to inject the HTML into the tag rather than using
`dangerouslySetInnerHTML`. Using JavaScript to set the HTML causes the browser
to fix any invalid markup and prevents it from breaking the renderer.

Fixes #1890 

## Testing

1. Visit the DSS flow, eg: http://calypso.localhost:3000/start/dss/
2. Click on a theme; Signup will move to step 2.
3. Click the "Back" link at the bottom of the page to return to step 1.
4. Verify the DSS page displays correctly (no sign of weird rendering).
5. Click on another theme.
6. Complete the signup flow.
7. Verify that the site is created and has the theme you chose.

aduth commented on client/components/dangerousHtml/index.jsx:
```diff
@@ -0,0 +1,28 @@
+
+export default React.createClass( {
+	propTypes: {
+		tag: React.PropTypes.string.isRequired,
```
Can we make this optional and default to `div` ?

sirbrillig replied:
That's a good idea. I actually kind of thought of that when I was writing the README.

aduth commented on client/components/dangerousHtml/index.jsx:
```diff
@@ -0,0 +1,28 @@
+
+	render() {
+		var tag = this.props.tag;
+		var props = {
```
Seems to be a good case to simply pass along any inherited props, except the ones that you have custom behavior for, e.g.:

```jsx
return React.createElement( tag, omit( this.props, 'tag', 'html ) );
```

(Means you could also lose the `className` `propType`)

aduth commented on client/components/dangerousHtml/index.jsx:
```diff
@@ -0,0 +1,28 @@
+
+	componentDidUpdate( prevProps ) {
+		if ( prevProps.html !== this.props.html ) {
+			this.componentDidMount();
```
We shouldn't call a lifecycle method like this. Instead, extract the logic to a function and call it from both places:

```jsx
componentDidMount() {
	this.updateHtml();
},

componentDidUpdate( prevProps ) {
	if ( prevProps.html !== this.props.html ) {
		this.updateHtml();
	}
},

updateHtml() {
	ReactDom.findDOMNode( this ).innerHTML = this.props.html;
}
```

aduth commented on client/components/dangerousHtml/index.jsx:
```diff
@@ -0,0 +1,28 @@
+import React from 'react';
```
Should have `External dependencies` docblock.

aduth replied:
Directory name should be `dangerous-html`, not `dangerousHtml`

aduth commented on client/components/dangerousHtml/index.jsx:
```diff
@@ -0,0 +1,28 @@
+export default React.createClass( {
+	propTypes: {
+		tag: React.PropTypes.string.isRequired,
+		html: React.PropTypes.string.isRequired,
```
We could probably default to an empty string.

aduth commented on client/signup/steps/dss/screenshot.jsx:
```diff
@@ -4,6 +4,8 @@
 import React from 'react';
 import classnames from 'classnames';
 
+import DangerousHtml from 'components/danger';
```
Bad import? Shouldn't this be `components/dangerous-html` ?

sirbrillig replied:
Oops. Yes. I changed the name and forgot to change this.

ehg commented:
I'm concerned that setting `innerHTML` directly will break React's tree conciliation pretty badly, which I've seen cause browser crashes before. Also, it's completely incompatible with server side rendering :(

aduth commented:
Discussed a bit privately, but let me reiterate here that anything related to dangerous HTML throws red flags in my book and should be avoided at all costs. Given that, I understand the need to be a bit more forgiving with the markup here.

I had mentioned the idea of using an `<iframe />` instead of dangerous HTML, and I think if this is possible, we should try to do so. See the [`<Shortcode />` component](https://github.com/Automattic/wp-calypso/tree/master/client/components/shortcode) as an example, which we could probably abstract to a `<SandboxedIframe />` component for rendering arbitrary markup. Specifically, see [`<ShortcodeFrame />`](https://github.com/Automattic/wp-calypso/blob/master/client/components/shortcode/frame.jsx).

Also, should we consider adopting similar mechanisms as used by Facebook to deter usage of this component? A verbose component name or prop indicating that it's poor-form to use?

sirbrillig commented:
> A verbose component name

Agreed, that was my intention with the name `DangerousHtml` ;-) But I'm open to other ideas.

sirbrillig commented:
> Also, it's completely incompatible with server side rendering

Hm. That's a good point. The other idea I had was to use some other module to fix invalid HTML. I'll look to see if something like `cheerio` can do it.

aduth commented:
>But I'm open to other ideas.

Maybe `dangerouslySetInnerHtml` as the prop name instead of `html` ? Redundant, yes, but that's kind of the point :smile:

ehg commented:
> Hm. That's a good point. The other idea I had was to use some other module to fix invalid HTML. 

We could do this on the server side, and store the valid markup in the theme CPT/ElasticSearch.

sirbrillig commented:
> Maybe dangerouslySetInnerHtml as the prop name

I changed it to `dangerousHtml` since I probably can't use the already-special `dangerouslySetInnerHtml`. I also added this to the README: 

**Warning**: This is still just as risky as using `dangerouslySetInnerHTML`
directly. Please avoid this component if at all possible.

sirbrillig commented:
Trying a different approach in #1903 by fixing the markup before it is rendered.

nylen commented:
> Please avoid this component if at all possible.

I agree with the README here :D

`dangerouslySetInnerHTML` is discouraged for good reasons, and I'd rather see us getting rid of existing uses than adding more ways to use it.

A couple of alternative approaches being discussed at #1903.
